src/main/kotlin/org/rust/lang/core/dfa/CFGBuilder.kt

```diff
@@ -603,7 +603,16 @@ class CFGBuilder(
 
         val prevGuards = ArrayDeque<CFGNode>()
 
-        matchExpr.arms.forEach { arm ->
+        val arms = matchExpr.arms
+        if (arms.isEmpty()) {
+            // The only case where a match with no arms is valid is when the discriminant is
+            // an empty type (!, empty enum etc), but the unreachability in this case is handled
```

Comment:
Good point. Note, IJ Rust does not produce unreachability warnings during type inference (in contrast to rustc). Instead, `CFGBuilder` has access to the already inferred and resolved types for each expression and can construct CFG accordingly. 

Empty enums are quite synthetic objects, but it looks easy to handle them here. Probably something like this will work:
```Kotlin
val discriminantEnum = (matchExpr.expr?.type as? TyAdt)?.item as? RsEnumItem
if (discriminantEnum == null || discriminantEnum.variants.isNotEmpty()) {
    addContainedEdge(discriminantExit, exprExit)
}
```
However, special handling of `!` type is not necessary indeed since such expressions are already covered (see `is TyNever` checks in `CFGBuilder`).

Reply:
We shouldn't special-case empty enums here since this is not the only case where the type is uninhabited. Other examples include a struct were one of the fields is uninhabited, an array or slice of uninhabited types, or an enum where each variant is uninhabited (these are quite easy to get with generics). Ideally there should be some sort of `isInhabited` query which depends on an early pass checking the structure of all types, and the unreachability lint would check it. Currently nothing like this is available, so empty enums are unhandled. This is a minor problem, since they are indeed quite rare, and false negatives are a much smaller issue than false positives.

Examples  of empty enums which are currently relatively often used are `std::convert::Infallible`, `byteorder::BigEndian/LittleEndian`, or various compile-time tokens. In fact, a seach with regexp `enum \w+ \s*\{\s*\}` in my projects and libraries gives 218 results, including 51 in the stdlib, so while empty enums aren't particularly common, they also aren't that rare.